File: app/src/main/java/com/salino/sali/data/repository/SuggestionsRepositoryImpl.kt
package com.salino.sali.data.repository

import com.salino.sali.data.local.source.RecurringLocalDataSource
import com.salino.sali.data.local.source.ShoppingLocalDataSource
import com.salino.sali.data.model.SuggestionItem
import com.salino.sali.domain.repository.SuggestionsRepository
import com.salino.sali.domain.service.SuggestionEngine
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SuggestionsRepositoryImpl @Inject constructor(
    private val shoppingLocalDataSource: ShoppingLocalDataSource,
    private val recurringLocalDataSource: RecurringLocalDataSource,
    private val suggestionEngine: SuggestionEngine
) : SuggestionsRepository {

    override fun observeSuggestions(householdId: String): Flow<List<SuggestionItem>> = combine(
        shoppingLocalDataSource.observeActiveItems(householdId),
        shoppingLocalDataSource.observeBoughtItems(householdId),
        recurringLocalDataSource.observeRecurringItems(householdId)
    ) { activeItems, boughtItems, recurringItems ->
        suggestionEngine.buildSuggestions(
            activeItems = activeItems,
            boughtItems = boughtItems,
            recurringItems = recurringItems
        )
    }
}

File: ios/Data/Repositories/SuggestionsRepositoryImpl.swift
import Foundation

@MainActor
final class SuggestionsRepositoryImpl: SuggestionsRepository {
    private let shoppingRepository: ShoppingRepository
    private let recurringRepository: RecurringRepository
    private let suggestionEngine: SuggestionEngine

    init(shoppingRepository: ShoppingRepository, recurringRepository: RecurringRepository, suggestionEngine: SuggestionEngine) {
        self.shoppingRepository = shoppingRepository
        self.recurringRepository = recurringRepository
        self.suggestionEngine = suggestionEngine
    }

    func observeSuggestions(householdId: String) -> AsyncStream<[SuggestionItem]> {
        let engine = suggestionEngine
        return AsyncStream { continuation in
            var activeItems: [ShoppingItem] = []
            var boughtItems: [ShoppingItem] = []
            var recurringItems: [RecurringItem] = []

            func yield() {
                continuation.yield(engine.buildSuggestions(
                    activeItems: activeItems,
                    boughtItems: boughtItems,
                    recurringItems: recurringItems,
                    now: Date()
                ))
            }

            let activeTask = Task { @MainActor in
                for await items in shoppingRepository.observeActiveItems(householdId: householdId) {
                    activeItems = items
                    yield()
                }
            }

            let boughtTask = Task { @MainActor in
                for await items in shoppingRepository.observeBoughtItems(householdId: householdId) {
                    boughtItems = items
                    yield()
                }
            }

            let recurringTask = Task { @MainActor in
                for await items in recurringRepository.observeRecurringItems(householdId: householdId) {
                    recurringItems = items
                    yield()
                }
            }

            continuation.onTermination = { _ in
                activeTask.cancel()
                boughtTask.cancel()
                recurringTask.cancel()
            }
        }
    }
}
